File: src/main/java/club/xiaojiawei/hearthstone/strategy/phase/DrawnInitCardPhaseStrategy.java
package club.xiaojiawei.hearthstone.strategy.phase;

import club.xiaojiawei.hearthstone.entity.Card;
import club.xiaojiawei.hearthstone.entity.ExtraEntity;
import club.xiaojiawei.hearthstone.entity.TagChangeEntity;
import club.xiaojiawei.hearthstone.enums.WarPhaseEnum;
import club.xiaojiawei.hearthstone.listen.PowerFileListen;
import club.xiaojiawei.hearthstone.status.War;
import club.xiaojiawei.hearthstone.strategy.PhaseStrategy;
import club.xiaojiawei.hearthstone.utils.PowerLogUtil;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import java.io.RandomAccessFile;

import static club.xiaojiawei.hearthstone.constant.GameConst.*;
import static club.xiaojiawei.hearthstone.constant.SystemConst.ROBOT;
import static club.xiaojiawei.hearthstone.enums.TagEnum.FIRST_PLAYER;

/**
 * @author 肖嘉威
 * @date 2022/11/27 13:35
 */
@Slf4j
public class DrawnInitCardPhaseStrategy extends PhaseStrategy {

    private String firstPlayerGameId;

    @SneakyThrows
    @Override
    public void dealing(String l) {
        War.setCurrentPhase(WarPhaseEnum.DRAWN_INIT_CARD_PHASE);
        log.info("当前处于：" + War.getCurrentPhase().getComment());
        RandomAccessFile accessFile = PowerFileListen.getAccessFile();
        ExtraEntity extraEntity;
        TagChangeEntity tagChangeEntity;
        while (true) {
            if ((l = accessFile.readLine()) == null){
                if (accessFile.getFilePointer() > accessFile.length() + 10000){
                    accessFile.seek(accessFile.length());
                }
                ROBOT.delay(1000);
            }else if (PowerFileListen.isRelevance(l)){
                PowerFileListen.setMark(System.currentTimeMillis());
                if (l.contains(SHOW_ENTITY)){
                    dealShowEntity(l, accessFile);
                }else if (l.contains(TAG_CHANGE)){
                    tagChangeEntity = PowerLogUtil.parseTagChange(l);
                    if (tagChangeEntity.getTag() == FIRST_PLAYER){
                        firstPlayerGameId = tagChangeEntity.getEntity();
                    }else {
                        PowerLogUtil.dealTagChange(tagChangeEntity);
                    }
                }else if (l.contains(FULL_ENTITY)){
                    extraEntity = PowerLogUtil.parseExtraEntity(l, accessFile);
                    extraEntity.setEntityName("幸运币");
                    Card card = new Card(extraEntity);
//                    向后手玩家手牌添加幸运币，确定后手玩家的gameId
                    switch (extraEntity.getPlayerId()){
                        case "1" -> {
                            War.getPlayer1().getHandArea().add(card);
                            War.getPlayer2().setGameId(firstPlayerGameId);
                        }
                        case "2" -> {
                            War.getPlayer2().getHandArea().add(card);
                            War.getPlayer1().setGameId(firstPlayerGameId);
                        }
                    }
                    firstPlayerGameId = null;
                }else if (l.contains("BLOCK_END")){
                    log.info(War.getCurrentPhase().getComment() + " -> 结束");
                    break;
                }
            }
        }
    }

    public void dealShowEntity(String l, RandomAccessFile accessFile){
        ExtraEntity extraEntity = PowerLogUtil.dealShowEntity(l, accessFile);
        if (War.getMe() == null) {
            switch (extraEntity.getPlayerId()) {
                case "1" -> {
                    War.setMe(War.getPlayer1());
                    War.setRival(War.getPlayer2());
                }
                case "2" -> {
                    War.setMe(War.getPlayer2());
                    War.setRival(War.getPlayer1());
                }
                default -> log.warn("不支持的playId");
            }
        }
    }
//    todo del
    public void dealTagChange(String l){
        String entityId = l.substring(l.indexOf("id") + 3, l.indexOf("zone=") - 1);
        if (War.getRival() == null){
            switch (l.substring(l.lastIndexOf("player") + 7, l.lastIndexOf("]"))){
                case "1" -> {
                    War.setRival(War.getPlayer1());
                    War.setMe(War.getPlayer2());
                }
                case "2" -> {
                    War.setRival(War.getPlayer2());
                    War.setMe(War.getPlayer1());
                }
                default -> log.warn("不支持的playId");
            }
        }
        Card card = War.getRival().getDeckArea().removeByEntityId(entityId);
        War.getRival().getHandArea().add(card);
        log.info("对方起始手牌之一：" + card);
    }

}

File: Hearthstone-Script-App/src/main/java/club/xiaojiawei/hsscript/strategy/phase/DrawnInitCardPhaseStrategy.kt
package club.xiaojiawei.hsscript.strategy.phase

import club.xiaojiawei.bean.Entity
import club.xiaojiawei.bean.isValid
import club.xiaojiawei.config.log
import club.xiaojiawei.data.COIN_CARD_ID
import club.xiaojiawei.enums.StepEnum
import club.xiaojiawei.enums.WarPhaseEnum
import club.xiaojiawei.hsscript.bean.log.ExtraEntity
import club.xiaojiawei.hsscript.bean.log.TagChangeEntity
import club.xiaojiawei.hsscript.enums.TagEnum
import club.xiaojiawei.hsscript.strategy.AbstractPhaseStrategy
import club.xiaojiawei.hsscript.utils.CardUtil

/**
 * 抽起始牌阶段
 * @author 肖嘉威
 * @date 2022/11/27 13:35
 */
object DrawnInitCardPhaseStrategy : AbstractPhaseStrategy() {

    /**
     * SHOW_ENTITY解析来的reverse为false
     * TAG_CHANGE解析来的reverse为true
     * @param playerId
     * @param reverse
     */
    fun verifyPlayer(playerId: String, reverse: Boolean) {
        var newPlayerId = playerId
        if (reverse) {
            newPlayerId = if (newPlayerId == "1") "2" else "1"
        }

        war.run {
            if (!me.isValid() && newPlayerId.isNotBlank()) {
                when (newPlayerId) {
                    "1" -> {
                        me = player1
                        rival = player2
                        log.info { "确定双方玩家号，我方1号，对方2号" }
                    }

                    "2" -> {
                        me = player2
                        rival = player1
                        log.info { "确定双方玩家号，我方2号，对方1号" }
                    }

                    else -> log.warn { "不支持的playId" }
                }
            }
        }
    }

    override fun dealTagChangeThenIsOver(line: String, tagChangeEntity: TagChangeEntity): Boolean {
        if (tagChangeEntity.tag == TagEnum.ZONE) {
            verifyPlayer(tagChangeEntity.playerId, true)
        } else if (tagChangeEntity.tag == TagEnum.NEXT_STEP && tagChangeEntity.value == StepEnum.BEGIN_MULLIGAN.name) {
            war.currentPhase = WarPhaseEnum.REPLACE_CARD
            return true
        }
        return false
    }

    override fun dealShowEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        if (Entity.isUnknownEntityName(extraEntity.entityName)) {
            verifyPlayer(extraEntity.playerId, false)
        }
        return false
    }

    /**
     * 确定一方玩家的游戏id，[.verifyPlayer]方法绝对会在此方法执行前执行
     * @param line
     * @param extraEntity
     * @return
     */
    override fun dealFullEntityThenIsOver(line: String, extraEntity: ExtraEntity): Boolean {
        val card = war.cardMap[extraEntity.entityId]
        card ?: let {
            log.warn { "card【entityId:${extraEntity.entityId}】不应为null" }
            return false
        }
        war.run {

            if (Entity.isUnknownEntityName(card.entityName) || card.entityName == "幸运币") {
                card.entityName = "幸运币"
                if (card.cardId.isNotBlank()) {
                    rival.gameId = firstPlayerGameId
                    log.info { "对方游戏id：$firstPlayerGameId" }
                } else {
                    me.gameId = firstPlayerGameId
                    log.info { "我方游戏id：$firstPlayerGameId" }
                }
                card.cardId = COIN_CARD_ID
                CardUtil.setCardAction(card)
                currentPlayer = if (me.gameId == firstPlayerGameId
                    || (rival.gameId.isNotBlank() && rival.gameId != firstPlayerGameId)
                ) {
                    me
                } else {
                    rival
                }
            }
        }
        return false
    }
}
